community/kernel/src/test/java/org/neo4j/kernel/impl/store/counts/CountsTrackerTest.java

```diff
@@ -267,9 +285,80 @@ public void shouldRotateOnDataChangesEvenIfTransactionIsUnchanged() throws Excep
         assertNotEquals( "rotated", before, tracker.currentFile() );
     }
 
+    @Test
+    @Resources.Life(STARTED)
+    public void shouldSupportTransactionsAppliedOutOfOrderOnRotation() throws Exception
+    {
+        // given
+        final CountsTracker tracker = the.managed( newTracker() );
+        try ( CountsAccessor.Updater tx = tracker.apply( 2 ).get() )
+        {
+            tx.incrementNodeCount( 1, 1 );
+        }
+        try ( CountsAccessor.Updater tx = tracker.apply( 4 ).get() )
+        {
+            tx.incrementNodeCount( 1, 1 );
+        }
+
+        // when
+        Future<Long> rotated = threading.executeAndAwait( new Rotation( 2 ), tracker, new Predicate<Thread>()
+        {
+            @Override
+            public boolean accept( Thread thread )
+            {
+                switch ( thread.getState() )
+                {
+                case BLOCKED:
+                case WAITING:
+                case TIMED_WAITING:
+                case TERMINATED:
+                    return true;
+                default:
+                    return false;
+                }
+            }
+        }, 10, SECONDS );
+        try ( CountsAccessor.Updater tx = tracker.apply( 5 ).get() )
+        {
+            tx.incrementNodeCount( 1, 1 );
+        }
+        try ( CountsAccessor.Updater tx = tracker.apply( 3 ).get() )
+        {
+            tx.incrementNodeCount( 1, 1 );
+        }
+
+        // then
+        assertEquals( "rotated transaction", 4, rotated.get().longValue() );
+        assertEquals( "stored transaction", 4, tracker.txId() );
+
+        // the value in memory
+        assertEquals( "count", 4, tracker.nodeCount( 1, Registers.newDoubleLongRegister() ).readSecond() );
+
+        // the value in the store
+        CountsVisitor visitor = mock( CountsVisitor.class );
+        tracker.visitFile( tracker.currentFile(), visitor );
+        verify( visitor ).visitNodeCount( 1, 3 );
+        verifyNoMoreInteractions( visitor );
+
+        assertEquals( "final rotation", 5, tracker.rotate( 5 ) );
+    }
+
     private CountsTracker newTracker()
     {
-        return new CountsTracker( the.logger(), the.fileSystem(), the.pageCache(), the.testPath() );
+        return new CountsTracker( the.logger(), the.fileSystem(), the.pageCache(), the.testPath() ).setInitializer(
+                new DataInitializer<CountsAccessor.Updater>()
+                {
+                    @Override
+                    public void initialize( CountsAccessor.Updater updater )
+                    {
+                    }
+
+                    @Override
+                    public long initialVersion()
+                    {
+                        return FileVersion.INITIAL_TX_ID;
+                    }
+                } );
```

Comment:
We could possibly set this "default initializer" in the constructor of the `CountsTracker`, but it doesn't appear to be needed in any case other than these tests.